kani-compiler/src/codegen_cprover_gotoc/reachability.rs

```diff
@@ -0,0 +1,586 @@
+                unreachable!("Unexpected at this MIR level")
+            }
+        }
+
+        self.super_terminator(terminator, location);
+    }
+}
+
+/// Visit every instruction in a function and collect the following:
+/// 1. Every function / method / closures that may be directly invoked.
+/// 2. Every function / method / closures that may have their address taken.
+/// 3. Every method that compose the impl of a trait for a given type when there's a conversion
+/// from the type to the trait.
+///    - I.e.: If we visit the following code:
+///      ```
+///      let var = MyType::new();
+///      let ptr : &dyn MyTrait = &var;
+///      ```
+///      We collect the entire implementation of `MyTrait` for `MyType`.
+/// 4. Every Static variable that is referenced in the function.
+/// 5. Drop glue? Static Initialization?
+///
+impl<'tcx> MirVisitor<'tcx> for MonoItemsCollector<'tcx> {}
+
+/// This is the reachability starting point. We start from every static item and proof harnesses.
+pub fn collect_reachable_items<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    starting_points: &[MonoItem<'tcx>],
+) -> FxHashSet<MonoItem<'tcx>> {
+    // For each harness, collect items using the same collector.
+    let mut collector = MonoItemsCollector {
+        tcx,
+        collected: FxHashSet::default(),
+        queue: vec![],
+        collected_sorted: vec![],
+    };
+    for item in starting_points {
+        collector.collect(*item);
+    }
+    collector.collected
+}
+
+/// Return whether we should include the item into codegen.
+/// We don't include foreign items only.
+fn should_codegen_locally<'tcx>(tcx: TyCtxt<'tcx>, instance: &Instance<'tcx>) -> bool {
+    if let Some(def_id) = instance.def.def_id_if_not_guaranteed_local_codegen() {
+        if tcx.is_foreign_item(def_id) {
+            // We cannot codegen foreign items.
+            false
+        } else {
+            // TODO: This should either be an assert or a warning.
+            // Need to compile std with --always-encode-mir first though.
+            // https://github.com/model-checking/kani/issues/1605
+            // assert!(tcx.is_mir_available(def_id), "no MIR available for {:?}", def_id);
+            //(!tcx.is_mir_available(def_id)).then(|| warn!(?def_id, "Missing MIR"));
+            tcx.is_mir_available(def_id)
+        }
+    } else {
+        // This will include things like VTableShim and other stuff. See the method
+        // def_id_if_not_guaranteed_local_codegen for the full list.
+        true
+    }
+}
+
+/// Extract the pair (concrete, trait) for a unsized cast.
+/// This function will return None if this is not a unsizing coercion from concrete to trait.
+///
+/// For example, if `&u8` is being converted to `&dyn Debug`, this method would return:
+/// `Some(u8, dyn Debug)`.
+///
+/// This method also handles nested cases and `std` smart pointers. E.g.:
+///
+/// Conversion between `Rc<Wrapper<T>>` into `Rc<Wrapper<dyn Debug>>` should return:
```

Comment:
Does this work on `T`? I thought it was concrete->trait

Reply:
T is a concrete type. I can improve this.